### ISSUE TITLE
Simplify requirements for Tekton's component images

### ISSUE THREAD

**Issue by dghubble** [tektoncd/pipeline#3060]
### Feature request

tekton-pipelines-controller arguments take a lengthy [list of images](https://github.com/tektoncd/pipeline/blob/master/config/controller.yaml#L58) for various helpers to support features across the breadth of Tekton's feature scope. I'm interested in the option to "scope this down" to `entrypoint` and `shell-image` (whether by optional images, disabling features, etc).

### Use case

Our use of Tekton is focused and minimalistic. Tasks and Pipeline's to run various container images in various orders. That's all. We have our own images with tools we need. If we need secrets, mount them as plain ordinary secrets. 

No use of PipelineResources (Git resource, PR resrouce, Image resource, cluser resource, storage resource), Tekton Authentication, image builders, or the various other quasi-helpers that seem built into Tekton. We've got that handled. Tekton tries to do a lot here (which I'm sure some appreciate), buts perhaps not everyone needs/wants that.

With that context, Tekton (as of v0.15.0) requires many images be [passed as arguments](https://github.com/tektoncd/pipeline/blob/master/config/controller.yaml#L58) (e.g. `found unset image flags`)

That list of images brings up concerns around auditing and compliance, whether there are code-paths / features we can disable if they're not core to our usage, and general questions around Tekton simplification and scope. For our use, we don't need Tekton to be able to pull and run its own images for `git`, `creds`, `imagedigest-exporter-image`, `pr-image`, `build-gcs-fetcher-image`, or `gsutil`.

In our typical usage, Tekton only uses the `entrypoint` and `shell-image` (init container)

* Can these various container images, their lineage, and their purpose be accounted for?
* Given we only need entrypoint and shell, can we as users disable features that would require the others?
* I recognize you face requests for more features/helpers every day. How can users provide some counter-weight to those making their way into core required arguments? Perhaps, "here are example images you might use in a pipeline on your own"

I'm hoping others share interest in simplification to this area.

**Comment by dlorenc** [comment]
Would it work for your use case to pass in invalid image names for the images you don't care about?

**Comment by dghubble** [comment]
Yes, it does help. Before v0.15, we removed the arguments (though apparently they had internal defaults https://github.com/tektoncd/pipeline/pull/3016) and with v0.15+ we can set them to empty string. The logic being if anyone tries to use something that would require them, it can fail (desirable in this case).

Perhaps one outcome from this issue becomes a docs PR. Even just enumerating these helpers aids in reasoning about what's required and (long-term) what belongs. 

Required:

| Name | Purpose |
|----------|------------|
| entrypoint-image | Used in initContainer for coordination of steps |
| shell-image         | Use in initContainer to place scripts                  |

Optional:

| Name | Purpose |
|----------|------------|
| kubeconfig-writer-image | ? |
| creds-image | Some helpers for auth?  |
| git-image      | Some helpers for git clone? |
| nop-image    | ? |
| imagedigest-exporter-image | ? |
| pr-image | ? |
| build-gcs-fetcher-image | ? |
| gsutil-image | Related to if you store temporary data in GCS |

**Comment by chmouel** [comment]
I think you would need to have the `nop-image` for the sidecar functionality, it's a dumb image with nothing less than an entrypoint that does nothing (and we actually used to have just `tianon/true` as the image there but that gets changed back to a binary) but help for our sidecar implementation.

**Comment by ImJasonH** [comment]
Setting the values to empty strings will result in controller crashes, but setting them to `"unset"` or any other image name should be fine.

Longer term PipelineResources should either be made to be pluggable and not bundled by default, or they should be phased out in favor of something else. I'm not sure which is the more likely future tbh.

**Comment by dghubble** [comment]
That sounds quite agreeable to me if they become pluggable or optional. So far I've not seen crashes, so maybe that validation isn't present in this release.

```
          image: gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/controller:v0.15.2
          args:
            - --namespace=tekton
            - --entrypoint-image=gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/entrypoint:v0.15.2
            - --shell-image=gcr.io/distroless/base:debug@sha256:f79e093f9ba639c957ee857b1ad57ae5046c328998bf8f72b30081db4d8edbe4
            # just to satisfy the validation
            - --kubeconfig-writer-image=""
            - --creds-image=""
            - --git-image=""
            - --nop-image=""
            - --imagedigest-exporter-image=""
            - --pr-image=""
            - --build-gcs-fetcher-image=""
            - --gsutil-image=""
```

**Comment by ImJasonH** [comment]
Weird, I'm able to trigger `CrashLoopBackOff` by passing `-creds-image=""`, or by not specifying the flag in controller.yaml. That change (#3016) was included in v0.15.

In any case, it sounds like the issue as stated is resolved -- you can pass some nonsensical value for images you know you don't care about, and things will work. Ideally we'd do a better job of documenting which images you don't need to care about given your usage, but even more ideally we'd drop the need for most of these optional built-in images anyway.

**Comment by dghubble** [comment]
Ok, it sounds like the direction is to drop most of the unneeded images and for now I can continue overriding them to be sure they're not used. I was hoping the documentation table could be filled out to have a barrier against more images being added, but sounds like that effort may not be needed / I'm not really using them to know what to put.

I'll close this out if there's no other concern
